Realm/ObjectServerTests/RLMAsymmetricSyncServerTests.mm

```diff
@@ -0,0 +1,316 @@
+}
+
+#pragma mark Asymmetric Sync App
+
+- (NSString *)asymmetricSyncAppId {
+    if (!_asymmetricSyncAppId) {
+        static NSString *s_appId;
+        if (s_appId) {
+            _asymmetricSyncAppId = s_appId;
+        }
+        else {
+            NSError *error;
+            NSArray *objectsSchema = @[[RLMObjectSchema schemaForObjectClass:PersonAsymmetric.class]];
+            _asymmetricSyncAppId = [RealmServer.shared createAppForAsymmetricSchema:objectsSchema error:&error];
+            if (error) {
+                NSLog(@"Failed to create asymmetric app: %@", error);
+                abort();
+            }
+            s_appId = _asymmetricSyncAppId;
+        }
+    }
+    return _asymmetricSyncAppId;
+}
+
+- (RLMApp *)asymmetricSyncApp {
+    if (!_asymmetricSyncApp) {
+        _asymmetricSyncApp = [RLMApp appWithId:self.asymmetricSyncAppId
+                                 configuration:self.defaultAppConfiguration
+                                 rootDirectory:self.clientDataRoot];
+        RLMSyncManager *syncManager = self.asymmetricSyncApp.syncManager;
+        syncManager.logLevel = RLMSyncLogLevelTrace;
+        syncManager.userAgent = self.name;
+    }
+    return _asymmetricSyncApp;
+}
+
+- (RLMUser *)userForSelector:(SEL)testSel {
+    return [self logInUserForCredentials:[self basicCredentialsWithName:NSStringFromSelector(testSel)
+                                                               register:YES
+                                                                    app:self.asymmetricSyncApp]
+                                     app:self.asymmetricSyncApp];
+}
+
+- (void)tearDown {
+    RLMUser *user = [self logInUserForCredentials:[RLMCredentials anonymousCredentials]
+                                              app:self.asymmetricSyncApp];
+    RLMMongoClient *client = [user mongoClientWithServiceName:@"mongodb1"];
+    RLMMongoDatabase *database = [client databaseWithName:@"test_data"];
+    RLMMongoCollection *collection = [database collectionWithName:@"PersonAsymmetric"];
+    [self cleanupRemoteDocuments:collection];
+    [super tearDown];
+}
+
+- (void)checkCountInMongo:(unsigned long)count {
+    RLMUser *user = [self logInUserForCredentials:[RLMCredentials anonymousCredentials]
+                                              app:self.asymmetricSyncApp];
+    RLMMongoClient *client = [user mongoClientWithServiceName:@"mongodb1"];
+    RLMMongoDatabase *database = [client databaseWithName:@"test_data"];
+    RLMMongoCollection *collection = [database collectionWithName:@"PersonAsymmetric"];
+    XCTestExpectation *findExpectation4 = [self expectationWithDescription:@"Should find documents"];
+    [collection findWhere:@{}
+               completion:^(NSArray<NSDictionary *> *documents, NSError *error) {
+        XCTAssertEqual(documents.count, count);
+        XCTAssertNil(error);
+        [findExpectation4 fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:60.0 handler:nil];
+}
```

Comment:
This is going to fail sometimes due to timing. Asymmetric objects use a fast-path on the server that bypasses the normal translator, but it still doesn't guarantee that the objects will be present in the mongo collection by the time the upload completes. We have to instead poll for completion and keep retrying until we get the expected count (with a maximum time, of course).

It also should use `countWhere:` rather than `findWhere:` to avoid pointlessly transferring and parsing the documents when we only need the count.